libminifi/include/core/ProcessGroup.h

```diff
@@ -205,6 +205,8 @@ class ProcessGroup : public CoreComponent {
    */
   std::shared_ptr<core::controller::ControllerServiceNode> findControllerService(const std::string &nodeId) const;
 
+  std::vector<std::shared_ptr<core::controller::ControllerServiceNode>> getAllControllerServices() const;
```

Comment:
Does this really need to give ownership to the callers?

Reply:
do you mean the vector should contain `const ControllerServiceNode*`s?

Reply:
That would be an improvement. Optionally, they could be wrapped in gsl::not_null. Whether a const refernce is enough for all consumers also needs to be considered, but if it is, then less unnecessary exposure is better. Currently it's pointers to mutable objects.

Reply:
I have changed this in 929aafecc10ae3d41ce8623d062de58c4f785abe

Now the interface is a bit of a mess, with a mix of return types; I have created a Jira to fix this: https://issues.apache.org/jira/browse/MINIFICPP-2312

Reply:
I'm assuming they can't be null, so I think it would be nicer to wrap the pointers in not_null. Or have runtime checks more graceful than gsl_Expects if they can be null.

Reply:
They can't be null, but I think `not_null` should be added to the container in `ControllerServiceMap`, so I would prefer to do this as part of MINIFICPP-2312.